### ISSUE TITLE
symbolic link in bash get wrong arg 0

### ISSUE THREAD

**Issue by SquallATF** [Alexpux/MSYS2-packages#1327]
a simple c++ source pring arg0
```cpp
#include <iostream>

int main(int argc, char const *argv[]) {
  std::cout << "argc: " << argc << std::endl;
  std::cout << "argv0: " << argv[0] << std::endl;
  return 0;
}
```
use follown comand compile it and make a symbolic link 
`g++ test.cpp -o test.exe && MSYS=winsymlinks:nativestrict ln -s test.exe testlink.exe`
then run `./testlink.exe` print follow result
```
argc: 1
argv0: E:\Work\cpptest\test.exe
```
but if run `testlink.exe` in cmd console will get right result
```
argc: 1
argv0: testlink.exe
```
in bash the  symbolic link arg0 will replace by real file location.

**Comment by SquallATF** [comment]
may be we don't need `PC_SYM_FOLLOW` when `perhaps_suffix` at [spawn.cc#L53](https://github.com/Alexpux/Cygwin/blob/091a0ac120e2c220a7b2c102e2e23d9b4b543ff6/winsup/cygwin/spawn.cc#L53), because we don't have cygwin style symlinks and don't need to get the real executable file name.

**Comment by SquallATF** [comment]
here is a temporary patch
```diff
From 56a25af0d3e741dc5877e2e6d6568ea56cd8f0c3 Mon Sep 17 00:00:00 2001
From: SquallATF <squallatf@gmail.com>
Date: Fri, 29 Jun 2018 20:22:44 +0800
Subject: [PATCH] fix native symbols link spawn passing wrong arg 0

---
 winsup/cygwin/path.cc  | 4 +++-
 winsup/cygwin/path.h   | 1 +
 winsup/cygwin/spawn.cc | 2 +-
 3 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/winsup/cygwin/path.cc b/winsup/cygwin/path.cc
index e5d5bcc69..15bf551e2 100644
--- a/winsup/cygwin/path.cc
+++ b/winsup/cygwin/path.cc
@@ -1021,7 +1021,9 @@ path_conv::check (const char *src, unsigned opt,
 		  if (component == 0 && !need_directory
 		      && (!(opt & PC_SYM_FOLLOW)
 			  || (is_rep_symlink ()
-			      && (opt & PC_SYM_NOFOLLOW_REP))))
+			      && (opt & PC_SYM_NOFOLLOW_REP))
+			  || ((sym.fileattr & FILE_ATTRIBUTE_REPARSE_POINT)
+			      && (opt & PC_SYM_NATIVE_FIX))))
 		    {
 		      /* last component of path is a symlink. */
 		      set_symlink (symlen);
diff --git a/winsup/cygwin/path.h b/winsup/cygwin/path.h
index 469d6076d..93f37917f 100644
--- a/winsup/cygwin/path.h
+++ b/winsup/cygwin/path.h
@@ -47,6 +47,7 @@ enum pathconv_arg
   PC_SYM_CONTENTS	= 0x0008,
   PC_NOFULL		= 0x0010,
   PC_NULLEMPTY		= 0x0020,
+  PC_SYM_NATIVE_FIX	= 0x0040,
   PC_POSIX		= 0x0080,
   PC_NOWARN		= 0x0100,
   PC_OPEN		= 0x0200,	/* use open semantics */
diff --git a/winsup/cygwin/spawn.cc b/winsup/cygwin/spawn.cc
index 264249072..2599c6a98 100644
--- a/winsup/cygwin/spawn.cc
+++ b/winsup/cygwin/spawn.cc
@@ -50,7 +50,7 @@ perhaps_suffix (const char *prog, path_conv& buf, int& err, unsigned opt)
 
   err = 0;
   debug_printf ("prog '%s'", prog);
-  buf.check (prog, PC_SYM_FOLLOW | PC_NULLEMPTY | PC_POSIX,
+  buf.check (prog, PC_SYM_FOLLOW | PC_SYM_NATIVE_FIX | PC_NULLEMPTY | PC_POSIX,
 	     (opt & FE_DLL) ? stat_suffixes : exe_suffixes);
 
   if (buf.isdir ())
-- 
2.17.1.windows.2
```

**Comment by awson** [comment]
Yeah, I've filed this back in #1134.

(btw, Microsoft since have fixed the similar WSL problem I mentioned there).

Dear MSys2 maintainers, please incorporate the fix proposed!

Otherwise this bugs renders MSys2 unusable in many scenarios, where the programs inspects it's own name to do things in a particular way (almost all llvm tools, various cross-compilers etc).

**Comment by SquallATF** [comment]
a new patch, forget the previous one
```diff
From 93d1049e4f8339cd6deea6c4cf3163588b601484 Mon Sep 17 00:00:00 2001
From: SquallATF <squallatf@gmail.com>
Date: Mon, 10 Sep 2018 17:45:27 +0800
Subject: [PATCH] fix native symbols link spwan wrong arg 0

---
 winsup/cygwin/spawn.cc | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/winsup/cygwin/spawn.cc b/winsup/cygwin/spawn.cc
index 264249072..6bcb01f51 100644
--- a/winsup/cygwin/spawn.cc
+++ b/winsup/cygwin/spawn.cc
@@ -50,7 +50,7 @@ perhaps_suffix (const char *prog, path_conv& buf, int& err, unsigned opt)
 
   err = 0;
   debug_printf ("prog '%s'", prog);
-  buf.check (prog, PC_SYM_FOLLOW | PC_NULLEMPTY | PC_POSIX,
+  buf.check (prog, PC_SYM_FOLLOW | PC_SYM_NOFOLLOW_REP | PC_NULLEMPTY | PC_POSIX,
 	     (opt & FE_DLL) ? stat_suffixes : exe_suffixes);
 
   if (buf.isdir ())
-- 
2.18.0.windows.1


```

**Comment by Alexpux** [comment]
@SquallATF what about defining ````PC_SYM_NOFOLLOW_REP````?

**Comment by SquallATF** [comment]
@Alexpux defined at [path.h#l46](https://cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=blob;f=winsup/cygwin/path.h;h=69954d8a58b91ae62e9089d951c4e94f7e84f752;hb=d83404e41808421442c8257392ecd095042cbfbf#l46),refer this commit [be37165](https://cygwin.com/git/gitweb.cgi?p=newlib-cygwin.git;a=commit;h=be371651146ca13155181a801d785cca77acbacb)

**Comment by Alexpux** [comment]
Fixed in 2.11.1-2 - https://github.com/Alexpux/MSYS2-packages/commit/e5eb8fa4a89c6370adf3f52692868c1e8247d816
